streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -628,15 +647,27 @@ public void onAssignment(final Assignment assignment) {
         Collections.sort(partitions, PARTITION_COMPARATOR);
 
         final AssignmentInfo info = AssignmentInfo.decode(assignment.userData());
-        final int usedVersion = info.version();
+        final int receivedAssignmentMetadataVersion = info.version();
+
+        if (usedSubscriptionMetadataVersion > receivedAssignmentMetadataVersion) {
```

Comment:
I think keeping it in nested condition is fine, even further, I think `if (receivedAssignmentMetadataVersion < 3)` should not happen since in this case the leader should have failed and not returning anything, so we can add an `else` to throw RTE to make it more strict.

Reply:
We cannot throw in `else` because this is a valid path during "manual" upgrade from version 1 or 2 to version 3.

In the second round of rolling bounces, the leader receives mixed subscriptions and sends old assignment back as long as at least one member sends and old subscription. Thus, instances that bounce the second time, will send new assignment and get old subscription back.